### PR TITLE
Full implementation; functionally equivalent to Rematch but better code

### PR DESCRIPTION
**Pull request by gafter**
Some differences from Rematch:
- Supports `::Type` as a pattern.
- Permits the use of already bound variables inside interpolations, e.g. the pattern `(x, y, $(x + 2))` matches the value `(1, 2, 3)`
- Binds types at macro expansion time, and generates assertions that the bindings are the same in the user's code at runtime.
- Produces (hopefully) excellent diagnostics when things go wrong
  - Errors point to a specific problem in the user's source
  - Exception stack traces will show the user code on the stack
  - Works with the debugger (points to your code instead of the Rematch macro implementation)
- Lowers a pattern-matching construct in several steps:
  - Binds source to an intermediate data structure that makes explicit the sequence of steps for matching
  - Performs some optimizations on that form
  - Lowers the intermediate form into code.

### CODE REVIEW

File: src/BindPattern.jl [hunk 1]
```diff
@@ -0,0 +1,288 @@
...
+        is_possible_type_name(t.args[2].value)
+end
+
+function bind_pattern!(
+    location::LineNumberNode,
+    source::Any,
+    input::Symbol,
+    state::BinderState,
+    assigned::ImmutableDict{Symbol,Symbol})
+    @assert !occursin("/src/", string(location.file))
```
Comment (nystrom) [inline src/BindPattern.jl:51]: This fails when I try to use `Rematch2` in another project.

File: src/Match.jl [hunk 2]
```diff
@@ -0,0 +1,111 @@
+function handle_match_eq(location::LineNumberNode, mod::Module, expr)
+    @capture(expr, pattern_ = value_) ||
+        error("Unrecognized match syntax: $expr")
+    input_variable::Symbol = gensym("input_value")
+    @assert !occursin("/src/", string(location.file))
```
Comment (nystrom) [inline src/Match.jl:5]: This fails when I try to use `Rematch2` in another project.

File: src/BindPattern.jl (hunk 1 above)
Comment (gafter) [reply src/BindPattern.jl]: Fixed.

File: README.md [hunk 3]
```diff
@@ -42,25 +42,31 @@ julia> f([1,2,3,4])
 (:at_least_2, 1, [2, 3], 4)
 
 julia> f([1])
-ERROR: Rematch2.MatchFailure([1])
+ERROR: MatchFailure([1])
 Stacktrace:
- [1] macro expansion at /home/jamie/.julia/v0.6/Rematch2/src/Rematch2.jl:173 [inlined]
- [2] f(::Array{Int64,1}) at ./REPL[3]:1
+ [1] f(x::Vector{Int64})
```
Comment (gafter) [inline README.md:47]: Diagnostics all now point to the problematic user code.

**Comment by Drvi** [comment]
I have mostly checked the generated code and played with it a bit and I had good time, this looks good :+1:. I tried comparing the performance to handwritten equivalents and didn't notice any overhead from the generated code.

I've macro-expanded the example from README and formatted it to make the review easier:
```julia
quote
    String == String || Rematch2.throw(Rematch2.AssertionError("String == String"))
    AbstractArray == AbstractArray || Rematch2.throw(Rematch2.AssertionError("AbstractArray == AbstractArray"))
    AbstractArray == AbstractArray || Rematch2.throw(Rematch2.AssertionError("AbstractArray == AbstractArray"))
    Foo == Foo || Rematch2.throw(Rematch2.AssertionError("Foo == Foo"))
    THE_INPUT = x
    if THE_INPUT isa String
        let;
            return :string 
        end
    elseif ((((((
        begin THE_INPUT isa AbstractArray end && 
            begin begin INPUT_LENGTH = Rematch2.length(THE_INPUT); true end end) && 
                begin Rematch2.isequal(INPUT_LENGTH, 3) end) && 
                    begin begin THE_INPUT_FIRST = Rematch2.getindex(THE_INPUT, 1); true end end) && 
                        begin begin THE_INPUT_2 = Rematch2.getindex(THE_INPUT, 2); true end end) && 
                            begin Rematch2.isequal(THE_INPUT_2, THE_INPUT_FIRST) end) && 
                                begin begin THE_INPUT_3 = Rematch2.getindex(THE_INPUT, 3); true end end) && 
                                    begin Rematch2.isequal(THE_INPUT_3, THE_INPUT_FIRST) end 
                                        let a = THE_INPUT_FIRST; 
                                            return (:all_the_same, a) 
                                        end
    elseif ((((
        begin THE_INPUT isa AbstractArray end && 
            begin begin INPUT_LENGTH = Rematch2.length(THE_INPUT); true end end) && 
                begin INPUT_LENGTH >= 2 end) && 
                    begin begin THE_INPUT_FIRST = Rematch2.getindex(THE_INPUT, 1); true end end) && 
                        begin begin THE_INPUT_MIDDLE = Rematch2.getindex(THE_INPUT, 2:Rematch2.length(THE_INPUT)-1); true end end) && 
                            begin begin THE_INPUT_LAST = Rematch2.getindex(THE_INPUT, Rematch2.length(THE_INPUT) + 0); true end end
                                let a = THE_INPUT_FIRST, bs = THE_INPUT_MIDDLE, c = THE_INPUT_LAST;
                                    return (:at_least_2, a, bs, c)
                                end
    elseif (((
        begin THE_INPUT isa Foo end && 
            begin begin THE_INPUT_X = Rematch2.getfield(THE_INPUT, :x); true end end) && 
                begin begin THE_INPUT_Y = Rematch2.getfield(THE_INPUT, :y); true end end) && 
                    begin Rematch2.isequal(THE_INPUT_Y, "foo") end) && 
                        begin let x = THE_INPUT_X; x > 1 end end
                            let x = THE_INPUT_X; 
                                return :foo 
                            end
    else
        Rematch2.throw(Rematch2.MatchFailure(THE_INPUT))
    end
end
```
Here are some notes:
* The assertion error messages at the top could maybe be more specific, like `Rematch2 code generation failed, Type1 != Type2`.
* When dealing with arrays, the generated code assumes one-based indexing which is fine for most cases, but to support packages like `OffsetArrays`, we should switch to `firstindex(a)` / `lastindex(a)` instead of `1` and `length(a)` etc.
* Since you check the array lengths first, you might use `@inbounds` when you touch the array (but I understand if you decided not to do it).   

>does it make sense to make this for Julia v1.8 and above only, so that I can test for the specific error messages produced when you do things wrong?

I saw people hiding some Julia-version-specific tests behind a `if VERSION >= ...`. I also saw people converting the tested output to string and using regexes to check the Julia-version-independent parts of the output.

File: src/BoundPattern.jl [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+        else
+            return new(location, source, gathered_subpatterns)
+        end
+    end
+end
+
+# Patterns which fetch intermediate values so they may be reused later without
+# being recomputed.  Each one has the side-effect of assigning a computed
+# value to a temporary variable.
+# abstract type Fetch <: BoundPattern
```
Comment (nystrom) [inline src/BoundPattern.jl:145]: Maybe move this comment down to the definition of `FetchPattern`.

**Comment by gafter** [comment]
I am going to merge this since neither the `main` branch or this branch has been reviewed.  It all needs to be reviewed.

**Pull request merged**